### PR TITLE
Passwords: Error detection and documentation

### PR DESCRIPTION
**Pull request by DerTanzschuh**
This introduces a mechanism to check curl commands for exit code `67` to catch authentication problems and displaying a proper error message in such cases.

In addition some test are adjusted and added to cover the new behaviour.

And then some documentation for password specific issues is added, in particular for passwords with special characters and the OSX keychain.

### CODE REVIEW

File: git-ftp [hunk 1]
```diff
@@ -622,7 +622,14 @@ set_deployed_sha1_failable() {
 
 set_deployed_sha1() {
 	set_deployed_sha1_failable
-	check_exit_status "Could not get last commit. Network down? Wrong URL? Use 'git ftp init' for the initial push." "$ERROR_DOWNLOAD"
+	local EXIT_CODE=$?
+	if [ "$EXIT_CODE" -eq 67 ]; then
+		print_error_and_die "Can't access remote 'ftp://test:***@127.0.0.1'. Failed to log in. Correct user and password? exiting..." "$ERROR_DOWNLOAD"
+	else
+		if [ "$EXIT_CODE" -ne 0 ]; then
```
Comment (mkllnk) [inline git-ftp]: I would probably go for a `case` statement here.

File: man/git-ftp.1.md [hunk 2]
```diff
@@ -559,7 +561,19 @@ hook. This hook is **not** bypassed by the --no-verify option.
...
+
+	--passwd '#my$fancy!secret'
+
+Mostly `--ask-passwd` works even if `--passwd` does not work. So maybe you can give this a try.
+
+Quoting also works if setting the [default](#defaults) with `git config`:
+
+	$ git config git-ftp.password '#my$fancy!secret'
+
+## netrc
```
Comment (mkllnk) [inline man/git-ftp.1.md]: Why did you change this to lower case?

File: git-ftp [hunk 3]
```diff
@@ -624,7 +624,7 @@ set_deployed_sha1() {
 	set_deployed_sha1_failable
 	local EXIT_CODE=$?
 	if [ "$EXIT_CODE" -eq 67 ]; then
-		print_error_and_die "Can't access remote 'ftp://test:***@127.0.0.1'. Failed to log in. Correct user and password? exiting..." "$ERROR_DOWNLOAD"
+		print_error_and_die "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$ERROR_DOWNLOAD"
```
Comment (mkllnk) [inline git-ftp]: This is a detour in the commit history. In the first commit I saw this mistake and almost commented on it. Then I thought that you may fix it in a later commit and kept on reading, keeping this in mind. It's easier for reviewers if you remove these detours with `git rebase -i` before submitting the pull request. Here are some guidelines I like: https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-commit#commits-should-not-make-changes-that-are-effectively-reverted-in-a-subsequent-commit

Ideally, I like to read a pull request like a book, page by page, commit by commit, each of them making sense and passing the test suite.

File: tests/git-ftp-test.sh [hunk 4]
```diff
@@ -151,6 +151,24 @@ test_init_fails() {
 	#assertEquals 1 $error_count
 }
 
+test_init_invalid_password() {
+	REMOTE_BASE_URL_DISPLAY="ftp://$GIT_FTP_USER:***@$GIT_FTP_HOST$GIT_FTP_PORT"
+	cd $GIT_PROJECT_PATH
```
Comment (mkllnk) [inline tests/git-ftp-test.sh]: This is already done in the setUp method. It's still in a lot of test methods from a time when it was not done in setUp. But we can actually remove them all (#spring-clean).

File: tests/git-ftp-test.sh [hunk 5]
```diff
@@ -151,6 +151,24 @@ test_init_fails() {
...
+	init="$($GIT_FTP_CMD init -u $GIT_FTP_USER -p wrongPassword $GIT_FTP_URL 2>&1)"
+	rtrn=$?
+	assertEquals "fatal: Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$init"
+	assertEquals 4 $rtrn
```
Comment (mkllnk) [inline tests/git-ftp-test.sh]: In one of my last commits I copied the error codes over to the test script. We can use them now which should improve the readability of the tests.

File: git-ftp [hunk 6]
```diff
@@ -498,7 +497,13 @@ upload_local_sha1() {
 	write_log "Uploading commit log to $REMOTE_BASE_URL_DISPLAY/${REMOTE_PATH}$DEPLOYED_SHA1_FILE."
 	if [ $DRY_RUN -ne 1 ]; then
 		echo "$LOCAL_SHA1" | upload_file - "$DEPLOYED_SHA1_FILE"
-		check_exit_status "Could not upload." "$ERROR_UPLOAD"
+		local EXIT_CODE=$?
+		if [ "$EXIT_CODE" -eq 67 ]; then
```
Comment (mkllnk) [inline git-ftp]: Is this now the third place with the same code? Maybe we should extract it into a function.

File: git-ftp [hunk 7]
```diff
@@ -1590,7 +1590,7 @@ do
 						REMOTE_PASSWD="$2"
 						shift
 					else
-						print_error_and_die "Too few arguments for option -p." "$ERROR_MISSING_ARGUMENTS"
+						print_error_and_die "Too few arguments for option -p. Maybe the manual will help: https://github.com/git-ftp/git-ftp/blob/master/man/git-ftp.1.md#passwords" "$ERROR_MISSING_ARGUMENTS"
```
Comment (mkllnk) [inline git-ftp]: Hm, this is a good idea but I'm worried about directing people to the master documentation even if people have a really old version. If Git-ftp is properly installed, the official way of reading the manual is `man git-ftp`. But maybe I'm a bit too old school and we should consider the master documentation valid for all versions (with version hints in there)?

I don't know.

File: man/git-ftp.1.md (hunk 2 above)
Comment (DerTanzschuh) [reply man/git-ftp.1.md]: In case you mean the headline … Because it’s an H2 now and not an H1 anymore.

File: git-ftp (hunk 3 above)
Comment (DerTanzschuh) [reply git-ftp]: Thank’s for the tip. I used `git rebase -i` in other cases before but did not think about it in this case. I’ll try to remember in the future. Thanks.

File: git-ftp (hunk 6 above)
Comment (DerTanzschuh) [reply git-ftp]: Yes. This a thought I also had.

I think a function like `check_exit_status` that process more than 2 parameters:
- status code variable (instead of `$?`)
- code to check for (instead of `-ne 0` use `-eq $2`)
- error message (as now)
- exit code to produce (as now)

Is it a idea to use exactly this function but alter the implementation if 4 parameters were given instead of 2 or should we create a new function?

I didn’t implement this cause It should be applied in some places as well I did not want change in this pull request.

File: tests/git-ftp-test.sh (hunk 5 above)
Comment (DerTanzschuh) [reply tests/git-ftp-test.sh]: Yes. I already was wondering why the codes are in the test script but are never used. So I decided to orientate myself on the existing test and do it the same way. I’d say this is a extra pull request since this should be applied to all test, shouldn’t it?

File: git-ftp (hunk 7 above)
Comment (DerTanzschuh) [reply git-ftp]: I see what you mean. I actually use both. But I also have no problem removing the hint. The suggestion was made [here](https://github.com/git-ftp/git-ftp/issues/468#issuecomment-501638166).

Comment (DerTanzschuh) [reply git-ftp]: Or we could make the link dynamic and use the current version to generate it. 

```sh
https://github.com/git-ftp/git-ftp/blob/$VERSION/man/git-ftp.1.md
```
And the anchor will also exit since it will be introduced in the version it is used. If a older version is installed the error message would not appear. And even if the anchor does not exits it will work since it is ignored then.

The only problem could be unreleased version, but I think people who use such a version could deal with it.

File: git-ftp (hunk 6 above)
Comment (DerTanzschuh) [reply git-ftp]: > I didn’t implement this cause It should be applied in some places as well I did not want change in this pull request.

Ok. Maybe I can introduce this here and we can use it in other case later.

File: man/git-ftp.1.md (hunk 2 above)
Comment (mkllnk) [reply man/git-ftp.1.md]: But other H2s are upper case as well. For example: `## RESET THE UPLOADED STATE`

File: tests/git-ftp-test.sh (hunk 5 above)
Comment (mkllnk) [reply tests/git-ftp-test.sh]: Yes, I agree.

File: git-ftp (hunk 6 above)
Comment (mkllnk) [reply git-ftp]: Interesting idea. So you would basically do this, right?
```diff
- [ "$EXIT_CODE" -eq 67 ] && print_error_and_die "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$ERROR_UPLOAD"
+ check_code_for_error "$EXIT_CODE" 67  "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$ERROR_UPLOAD"
```

It doesn't improve the code much and we still have the repetition. I think we need a method to check curl exit codes:

```sh
check_curl_status() {
  local curl_exit_status=$?
  local unknown_error_message=$1
  local error_code=$2
  case $? in
    0) ;;
    67) print_error_and_die "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$error_code" ;;
    *) print_error_and_die "Could not upload." "$error_code"
}
```

Maybe that's not the best code yet. You probably get the idea. Maybe we need first a function that maps the curl error code to a good error message and if it returns an error message we then exit:

```sh
message="$(check_curl_status $?)"
if [ -n "$message" ]; then
  print_error_and_die "Could not get last commit: $message exiting..." $ERROR_DOWNLOAD"
fi
```

File: git-ftp (hunk 7 above)
Comment (mkllnk) [reply git-ftp]: I'm happy with the link for now. None of the other solutions convince me. :wink:

Comment (DerTanzschuh) [reply git-ftp]: Too bad … 😉

The dynamic link with the version or the static one to master?

File: man/git-ftp.1.md (hunk 2 above)
Comment (DerTanzschuh) [reply man/git-ftp.1.md]: Douh 😆

File: git-ftp (hunk 7 above)
Comment (mkllnk) [reply git-ftp]: master

File: git-ftp (hunk 6 above)
Comment (DerTanzschuh) [reply git-ftp]: Yes. Except for the function name, this is exactly my idea.

Oh. yes. I got the point. I like your idea, but it would not be that flexible. The function `check_code_for_error` can be used in other cases than curl where we have to check for exit codes. 

But we could do one thing and not lose the other.

And the idea with the message mapping is great as well, but I’d do it the other way round. It could look like this:

```sh
check_curl_status $? "Could not get last commit"

# Checks if a given curl exit code was successful and if not exits with a nice error message (in some cases)
#
# $1 - exit code to check
# $2 - message: a custom message that is put in front of the error message
# $3 - error code: code to exit with if $1 is bad
#
check_curl_status() {
  case $1 in
    0) ;;
    67) print_error_and_die "$2: Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$3" ;;
    *) print_error_and_die "$2" "$3"
}
```
This way the code where the check is called becomes cleaner and we don’t need an additional if there. In addition this would make this function similar to the existing behaviour of `check_exit_status` since this exits if the status is bad, too.

And in addition to this we can also provide the other function(s) to use it in other case than curl where we need a more flexible approach:
```sh
# Checks if a given exit code equals a given value
#
# $1 - exit code to check
# $2 - code to test against
# $3 - error message
# $4 - error code to produce
#
check_exit_status_equals() {
	if [ $1 -eq $2 ]; then
		print_error_and_die "$3, exiting..." "$4"
	fi
}

# Checks if a given exit code not equals a given value
#
# $1 - exit code to check
# $2 - code to test against
# $3 - error message
# $4 - error code to produce
#
check_exit_status_not_equals() {
	if [ $1 -ne $2 ]; then
		print_error_and_die "$3, exiting..." "$4"
	fi
}
```

Comment (mkllnk) [reply git-ftp]: Yes, great. Go ahead!

**Comment by DerTanzschuh** [comment]
I hope I watched all the comment and didn’t miss anything.

I will do the error codes and old `cd`s (i removed the ones in my new tests) in tests in another PR.
